### PR TITLE
str #20262 reduce should fail explicitly on empty stream

### LINKED ISSUE #20262: Reduce emits null if used with AnyRef and an empty source
Reduce initially uses a default value as aggregate:

```Scala
var aggregator: T = _
```

Which means that it is `null` for non-primitives. If upstream closes before emitting any element the `aggregator` value is emitted downstream that emits a `null` downstream which is illegal.

I think it should just close rather than pass the empty value downstream, if you want a seed that is emitted on empty upstream you can always use `fold` instead.

### PR DESCRIPTION
**Pull request by ktoso**
Was a fun fix for a change :-)
Resolves https://github.com/akka/akka/issues/20262

### CODE REVIEW

**Comment by viktorklang** [comment]
@ktoso oops, didn't see you already tackled this. Ohwell, please feel encouraged to steal my performance improvemens from https://github.com/akka/akka/pull/20268

**Comment by ktoso** [comment]
No problem :) thanks for the opt, I'll steal it into here :+1:

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -1168,6 +1168,9 @@ private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraph
         pull(in)
         setHandler(in, rest)
       }
+      override def onUpstreamFinish(): Unit = {
+        failStage(new NoSuchElementException("reduce over empty stream"))
+      }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: So simple fix it's almost suspicious ;)

**Comment by johanandren** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Yep, now it only needs the opts from my PR :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 2]
```diff
@@ -107,6 +107,11 @@ final class Source[+Out, +Mat](private[stream] override val module: Module)
    * The returned [[scala.concurrent.Future]] will be completed with value of the final
    * function evaluation when the input stream ends, or completed with `Failure`
    * if there is a failure signaled in the stream.
+   *
+   * If the stream is empty (i.e. completes before signalling any elements),
+   * the reduce stage will fail its downstream with an [[NoSuchElementException]],
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: s/an/a/

**Comment by patriknw** [comment]
LGTM

**Comment by ktoso** [comment]
Will merge after validation

**Pull request merged**